Openworld: Convert room_generate() into a general purpose, single-room loading + connecting function

Justin08784 opened the pull request:
- `room_generate()` was refactored into a general-purpose function for generating a single room around any given room in the game, and connecting the two rooms with paths. We did this by adding `room_t *curr`, `char *dir_to_curr`, and `char *dir_to_new` params. We removed the `gencontext_t *context` param because it served no purpose (as of now).
- redundant code in recursive_generate() and multi_room_generate() was replaced with a call to room_generate().
- room_generate() functionality for picking random directions was organized into a new helper called pick_random_dir(). Relevant tests were added to test_autogenerate.c.
- some bugs were fixed (details [here](https://github.com/uchicago-cs/chiventure/issues/1033#issuecomment-846226439))

ahmed-sumaiya reviewed (approved):
Looks good to me!

MaxineK36 commented on include/openworld/autogenerate.h:
```diff
@@ -58,25 +58,48 @@ bool path_exists_in_dir(room_t *r, char *direction);
+ * - FAILURE: otherwise
+ */
+
+int pick_random_dir(room_t *curr, char *out_dir_to_curr, char *out_dir_to_new);
```
Since `dir` is so often shortened to `directory`, it would be helpful to call this `pick_random_direction` (and change all other abbreviations)

MaxineK36 commented on include/openworld/autogenerate.h:
```diff
@@ -246,12 +269,12 @@ int multi_room_level_generate(game_t *game, gencontext_t *context,
  * - game: pointer to a game struct. Should not be NULL
  * - room_t *curr_room: pointer to the room to serve as the pivot
- * - speclist_t *speclist: the llist of roomspect_t that each hold info for a separate room
+ * - gencontex_t *context: pointer to a gencontext. Not NULL.
```
By "Not NULL" do you mean it's guaranteed somehow not to be null, or expected not to be null? this should be specified

Justin08784 replied:
Should not be.

MaxineK36 commented on include/openworld/autogenerate.h:
```diff
@@ -246,12 +269,12 @@ int multi_room_level_generate(game_t *game, gencontext_t *context,
  *               (must be a subset of the default six: "NORTH", "EAST", "SOUTH", "WEST", "UP", "DOWN")
  * - num_of_dir: array length of directions, i.e. number of directions
- * - dir_to_parent: direction to the parent pivot
+ * - dir_to_parent: direction to the parent pivot. Not NULL.
```
ditto here, and please change `dir`

MaxineK36 commented on src/openworld/src/autogenerate.c:
```diff
@@ -56,8 +56,9 @@ room_t* roomspec_to_room(roomspec_t *roomspec)
+int pick_random_dir(room_t *curr, char *out_dir_to_curr, char *out_dir_to_new)
 {
     /* 2D array of possible directions */
     char directions[4][6];
```
This `4` should be a defined constant, and then should be used as such later. e.g.
`int NUM_DIRECTIONS = 4` (or, at the beginning of the file, `#define NUM_DIRECTIONS 4`) and then you should use that constant throughout

Justin08784 replied:
Defined constant NUM_COMPASS_DIRECTIONS 4 at the top of file

MaxineK36 commented on src/openworld/src/autogenerate.c:
```diff
@@ -67,37 +68,49 @@ int room_generate(game_t *game, gencontext_t *context, roomspec_t *rspec)
 
     /* Random initial direction */
-    unsigned int first_direction = rand() % 4;
+    unsigned int initial_dir = rand() % 4;
```
abbreviations

MaxineK36 commented on src/openworld/src/autogenerate.c:
```diff
@@ -67,37 +68,49 @@ int room_generate(game_t *game, gencontext_t *context, roomspec_t *rspec)
+    /* create new combination of rooms/items from randomly picked roomspec
+    Adds one generated room from the head of context->speclist only */
+    room_t *new_room = roomspec_to_room(rspec_new);
+    assert(SUCCESS == add_room_to_game(game, new_room)); // this step should never fail
```
why should it never fail?

Justin08784 replied:
room_generate() needs to add a room to the game. If this fails, it should fail.

MaxineK36 commented on src/openworld/src/autogenerate.c:
```diff
@@ -67,37 +68,49 @@ int room_generate(game_t *game, gencontext_t *context, roomspec_t *rspec)
+
+    /* Path to the generated room */
+    path_t* path_to_new = path_new(new_room, dir_to_new);
+    if (add_path_to_room(curr, path_to_new)) // but this step can? 
```
why?

MaxineK36 replied:
Also is that meant to be `if (add_path_to_room(curr, path_to_new) == SUCCESS)`?

Justin08784 replied:
swapped if statements these with asserts

MaxineK36 commented on src/openworld/src/autogenerate.c:
```diff
@@ -67,37 +68,49 @@ int room_generate(game_t *game, gencontext_t *context, roomspec_t *rspec)
+    path_t* path_to_new = path_new(new_room, dir_to_new);
+    if (add_path_to_room(curr, path_to_new)) // but this step can? 
+        return FAILURE;
+    /* should we not make it a responsibility of the caller 
```
What did you determine here? These comments should be removed before merging.

Justin08784 replied:
I have decided to make it the caller's responsibility, comment has been removed.

MaxineK36 commented on src/openworld/src/autogenerate.c:
```diff
@@ -108,18 +121,25 @@ int multi_room_generate(game_t *game, gencontext_t *context, char *room_id, int
-        room_generate(game, context, rspec);
+
+        int rc;
+        char dir_to_curr[6], dir_to_new[6];
```
where is this 6 coming from?

Justin08784 replied:
Defined constant NUM_DIRECTIONS 6

there are six default directions: the compass points + up + down

MaxineK36 commented on src/openworld/src/autogenerate.c:
```diff
@@ -108,18 +121,25 @@ int multi_room_generate(game_t *game, gencontext_t *context, char *room_id, int
+        char dir_to_curr[6], dir_to_new[6];
+
+        rc = pick_random_dir(game->curr_room, dir_to_curr, dir_to_new);
+        if (rc == FAILURE) 
```
You don't need an intermediate integer for this, since you only use it once.

MaxineK36 commented on src/openworld/src/autogenerate.c:
```diff
@@ -160,14 +180,18 @@ item_hash_t *random_items(roomspec_t *room)
     }
```
This function doesn't seem to appear in autogenerate.h

Justin08784 replied:
The function is in autogenerate.h. idk

MaxineK36 reviewed (changes requested):
A couple of notes; the biggest one is that `dir` is frequently used to abbreviate `directory`, and so to avoid confusion `direction` should be written out

Justin08784 commented:
- Removed dir abbrevs
- Removed magic numbers; introduced constants MAX_DIRECTION_STRLEN, NUM_DIRECTIONS, NUM_COMPASS_DIRECTIONS
- Changed documentation "Not NULL." to "Should not be NULL."

Justin08784 commented:
- Replaced 'if' statements in room_generate() with asserts. Made it caller's responsibility to ensure parameter `direction_to_new` is an available direction; if it is not available, an assert will crash room_generate (since this also means that room_generate_failure test is not needed, it was deleted)
-

MaxineK36 reviewed (changes requested):
One last nitpick, otherwise looks great!

MaxineK36 reviewed (approved):
Looks great!

Justin08784 commented:
Closing statement:
Comments by reviewers have been addressed and branch `openworld/refactor-room-generate` is now merged into dev. 

room_generate() was refactored into a general-purpose function for generating a single room around any given room in the game, and connecting the two rooms with paths. It is now called by both higher-level generation functions multi_room_generate() and recursive_generate().

room_generate()'s functionality for picking random directions was organized into a new helper called pick_random_direction(). Relevant tests were added to test_autogenerate.c.

Followups:
The dir abbreviation has been removed and several functions were changed. The source document may need to be updated.

MaxineK36 commented:
Issue Score: ✔️++

Comments:
Excellent work! Your summary, updates, and closing statement are all particularly detailed and well-written.

<!-- {"score": "check-plus-plus"} -->
